qiskit/circuit/bit.py

```diff
@@ -49,33 +54,50 @@ def _update_hash(self):
     @property
     def register(self):
         """Get bit's register."""
+        if (self._register, self._index) == (None, None):
+            raise CircuitError('Attmped to query register of a new-style Bit.')
+
         return self._register
 
     @register.setter
     def register(self, value):
         """Set bit's register."""
+        if (self._register, self._index) == (None, None):
+            raise CircuitError('Attmped to set register of a new-style Bit.')
+
         self._register = value
         self._update_hash()
 
     @property
     def index(self):
         """Get bit's index."""
+        if (self._register, self._index) == (None, None):
+            raise CircuitError('Attmped to query index of a new-style Bit.')
+
         return self._index
 
     @index.setter
     def index(self, value):
         """Set bit's index."""
+        if (self._register, self._index) == (None, None):
+            raise CircuitError('Attmped to query index of a new-style Bit.')
+
         self._index = value
         self._update_hash()
 
     def __repr__(self):
         """Return the official string representing the bit."""
+        if (self._register, self._index) == (None, None):
+            return "%s(%s)" % (self.__class__.__name__, hex(id(self)))
```

Comment:
Is there a reason not to cache this? It definitely will be faster than the repr based on the register case, but we still end up using this a lot (because we use str(qargs) as the sort key for `lexicographical_topological_sort()`) and this basically gets called once for each bit in an op's qargs list each time we create a new DAGNode. So I wonder what the overhead is going to be for doing the string manipulation and hex conversion once per op vs once per bit.

Reply:
This is another sidestep to have new-style `Bit`s look more like placeholder objects (though looking now, similarly using `object.__str__` would've been better). I'll update this.

For `lexicographical_topological_sort()`, I was hoping to remove the string comparison because we would only need an object comparison with new-style `Bit`s, but looking again, `rx.lexicographical_topological_sort` expects a `key` function that returns a string. I missed this in the DAGCircuit PR so this is a good catch, I'll update `DAGNode.sort_key` there and if string construction still shows up in profiling, we can resume caching.

Reply:
Yeah doing the string comparison on the retworkx rust side was done for performance. When we first added that function about a year ago having the callback return a string key was a conscious tradeoff between covering a wider set of use cases and performance for how DAGCircuit used it because at least back then the testing showed rust side string comparison to be faster than a python callback returning a bool, but this might have changed some in the past year).